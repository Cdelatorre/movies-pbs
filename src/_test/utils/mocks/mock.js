export const moviesData = {
  viewedMovies: [
    {
      id: "1",
      title: "Movie 1",
      genres: ["Horror", "Comedy"],
      viewed: true,
      img: "https://picsum.photos/id/118/300/400",
      createdAt: "2022-10-16T20:25:58.420Z",
    },
  ],
  unViewedMovies: [
    {
      id: "2",
      title: "Movie 2",
      genres: ["Horror", "Comedy"],
      viewed: false,
      img: "https://picsum.photos/id/159/300/400",
      createdAt: "2022-10-16T20:51:29.081Z",
    },
    {
      id: "3",
      title: "Movie 3",
      genres: ["Romance"],
      viewed: false,
      img: "https://picsum.photos/id/122/300/400",
      createdAt: "2022-10-16T20:51:45.249Z",
    },
    {
      id: "4",
      title: "Movie 4",
      genres: ["Test"],
      viewed: false,
      img: "https://picsum.photos/id/168/300/400",
      createdAt: "2022-10-16T20:52:14.784Z",
    },
    {
      id: "5",
      title: "Movie 5",
      genres: ["Random"],
      viewed: false,
      img: "https://picsum.photos/id/149/300/400",
      createdAt: "2022-10-16T20:52:36.529Z",
    },
  ],
};

export const singleMovieData = {
  id: "1",
  title: "Movie 1",
  genres: ["Horror", "Comedy", "Romance"],
  viewed: false,
  img: "https://picsum.photos/id/149/300/400",
  createdAt: "2022-10-16T20:52:36.529Z",
};

export const activeFiltersData = {
  activeFilters: [],
  search: "",
};

export const activeLoadingData = {
  value: false,
};

export const emptyStoreData = {
  movies: { viewedMovies: [], unViewedMovies: [] },
  filters: { activeFilters: [], search: "" },
  loading: { value: false },
};

export const sampleFiltersData = ["Horror", "Comedy", "Romance"];

export const customStoreData = (movies, activeFilters, loading) => {
  return {
    movies: movies ? moviesData : emptyStoreData.movies,
    filters: activeFilters ? activeFiltersData : emptyStoreData.filters,
    loading: loading ? activeLoadingData : emptyStoreData.loading,
  };
};
