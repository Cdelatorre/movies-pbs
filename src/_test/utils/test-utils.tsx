import React, { PropsWithChildren } from "react";
import { render } from "@testing-library/react";
import type { RenderOptions } from "@testing-library/react";
import type { PreloadedState } from "@reduxjs/toolkit";
import { Provider } from "react-redux";
import { AppStore, RootState, setupStore } from "../../store";
import { BrowserRouter, Router } from "react-router-dom";
import { emptyStoreData } from "./mocks/mock";

interface ExtendedRenderOptions extends Omit<RenderOptions, "queries"> {
  preloadedState?: PreloadedState<RootState>;
  store?: AppStore;
}

export function renderWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = emptyStoreData,

    store = setupStore(preloadedState),
    ...renderOptions
  }: ExtendedRenderOptions = {},
  route?: string
) {
  function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
    if (route) window.history.pushState({}, "Test routing", route);

    return (
      <BrowserRouter>
        <Provider store={store}>{children}</Provider>;
      </BrowserRouter>
    );
  }
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}

export function renderWithRouter(ui: React.ReactElement, history?: any) {
  function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
    if (history) {
      return (
        <Router location="/" navigator={history}>
          {children}
        </Router>
      );
    } else {
      return <BrowserRouter>{children}</BrowserRouter>;
    }
  }
  return { ...render(ui, { wrapper: Wrapper }) };
}
