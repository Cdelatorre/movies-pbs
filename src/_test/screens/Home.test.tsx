import { fireEvent, screen } from "@testing-library/react";
import Home from "../../screens/home";
import { renderWithProviders } from "../utils/test-utils";
import { customStoreData } from "../utils/mocks/mock";

const customStore = customStoreData(true, false, false);
const movies = [
  ...customStore.movies.unViewedMovies,
  ...customStore.movies.viewedMovies,
];

describe("Renders Home page correctly", () => {
  test("Renders Each component correctly", () => {
    renderWithProviders(<Home />);

    const homeScreen = screen.getByTestId("home-component");
    const form = screen.getByTestId("form-component");
    const filters = screen.getByTestId("filters-component");
    const movies = screen.getByTestId("movies-list");

    expect(homeScreen).toBeInTheDocument();
    expect(form).toBeInTheDocument();
    expect(filters).toBeInTheDocument();
    expect(movies).toBeInTheDocument();

    expect(
      screen.getByRole("checkbox", { name: "Horror" })
    ).toBeInTheDocument();
    expect(
      screen.getByRole("checkbox", { name: "Comedy" })
    ).toBeInTheDocument();
    expect(
      screen.getByRole("heading", {
        name: "Vaya, no hemos encontrado pelÃ­culas...",
      })
    ).toBeInTheDocument();

    expect(homeScreen).toMatchSnapshot();
  });

  test("Renders Movie component for each card", () => {
    renderWithProviders(<Home />, {
      preloadedState: {
        ...customStore,
      },
    });

    movies.forEach((movie) => {
      expect(screen.getByTestId(`movie-card-${movie.id}`)).toBeInTheDocument();
    });
  });

  test("Renders Movie component with specific info", () => {
    renderWithProviders(<Home />, {
      preloadedState: {
        ...customStore,
      },
    });

    customStore.movies.unViewedMovies.forEach((movie) => {
      expect(screen.getByText(movie.title)).toBeInTheDocument();
    });
  });

  test("Renders Skeletons component when deleting a movie", () => {
    renderWithProviders(<Home />, {
      preloadedState: {
        ...customStore,
      },
    });

    const deleteButton = screen.getByTestId("delete-movie-btn-1");
    fireEvent.click(deleteButton);

    expect(screen.queryByTestId("movie-skeleton-1")).toBeNull();
    expect(screen.getByTestId("movie-skeleton-2")).toBeInTheDocument();
    expect(screen.getByTestId("movie-skeleton-3")).toBeInTheDocument();
    expect(screen.getByTestId("movie-skeleton-4")).toBeInTheDocument();
    expect(screen.getByTestId("movie-skeleton-5")).toBeInTheDocument();
  });
});
