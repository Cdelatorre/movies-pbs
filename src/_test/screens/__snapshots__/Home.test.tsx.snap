// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders Home page correctly Renders Each component correctly 1`] = `
<div
  data-testid="home-component"
>
  <div
    class="mt-5"
  >
    <div
      class="container d-flex justify-content-center"
    >
      <div
        class="movie-form-container mt-3"
        data-testid="form-component"
      >
        <h2
          class="movie-form-title mt-0"
        >
          Create your movie
        </h2>
        <form
          class="movie-form"
          data-testid="movie-form"
        >
          <div
            class="input "
          >
            <label
              class="form-label w-100 text-start text-light"
              for="title-input"
            >
              Title
            </label>
            <input
              class="form-control form-text bg-dark text-light"
              data-testid="input-text-title"
              id="title-input"
              name="title"
              placeholder="Write down a title for your movie"
              type="text"
              value=""
            />
          </div>
          <div
            class="genders d-flex mb-3 mt-3"
          />
          <div
            class="input "
          >
            <label
              class="form-label w-100 text-start text-light"
              for="title-input"
            >
              Genres
            </label>
            <input
              class="form-control form-text bg-dark text-light"
              data-testid="input-text-currentGenre"
              id="title-input"
              name="currentGenre"
              placeholder="Write down some movie genres"
              type="text"
              value=""
            />
          </div>
        </form>
        <div
          class="mt-4"
        >
          <button
            class="btn btn-success w-100 mt-3"
            data-testid="test-submit-form-btn"
            id="submit-form-btn"
          >
            Create
          </button>
        </div>
      </div>
    </div>
    <div
      class="container"
    >
      <div
        class="filters-container d-flex flex-column align-items-start"
        data-testid="filters-component"
      >
        <h2
          class="text-light text-start mb-4"
        >
          Find your movie!
        </h2>
        <div
          class="sub-decoration"
        />
        <div
          class="d-flex flex-wrap"
        >
          <div
            class="d-flex text-light flex-row flex-wrap w-100"
          >
            <div
              class="input form-check"
            >
              <label
                class="d-inline ps-2 pe-3 text-start text-light"
                for="Horror"
              >
                Horror
              </label>
              <input
                class="form-control form-checkbox bg-dark text-light"
                data-testid="input-checkbox-Horror"
                id="Horror"
                name="Horror"
                readonly=""
                type="checkbox"
                value="Horror"
              />
            </div>
            <div
              class="input form-check"
            >
              <label
                class="d-inline ps-2 pe-3 text-start text-light"
                for="Romance"
              >
                Romance
              </label>
              <input
                class="form-control form-checkbox bg-dark text-light"
                data-testid="input-checkbox-Romance"
                id="Romance"
                name="Romance"
                readonly=""
                type="checkbox"
                value="Romance"
              />
            </div>
            <div
              class="input form-check"
            >
              <label
                class="d-inline ps-2 pe-3 text-start text-light"
                for="Comedy"
              >
                Comedy
              </label>
              <input
                class="form-control form-checkbox bg-dark text-light"
                data-testid="input-checkbox-Comedy"
                id="Comedy"
                name="Comedy"
                readonly=""
                type="checkbox"
                value="Comedy"
              />
            </div>
            <button
              class="btn btn-primary"
              data-testid="test-reset-filters-btn"
              id="reset-filters-btn"
            >
              <i
                class="fas fa-sync-alt"
              />
            </button>
          </div>
        </div>
        <div
          class="search-bar mb-5 mt-4 w-100 d-flex"
          data-testid="search-bar"
        >
          <div
            class="input "
          >
            <input
              class="form-control form-text bg-dark text-light w-100"
              data-testid="input-text-search"
              name="search"
              placeholder="Search movie"
              type="text"
              value=""
            />
          </div>
          <button
            class="btn btn-success ms-3"
            data-testid="test-search-btn"
            id="search-btn"
          >
            Search
          </button>
        </div>
      </div>
    </div>
    <div
      class="container movie-list"
      data-testid="movies-list"
    >
      <div
        class="row mb-5"
      >
        <div
          class="container"
        >
          <h4
            class="text-light mb-5 p-5 border rounded border-light"
          >
            Vaya, no hemos encontrado pel√≠culas...
          </h4>
        </div>
      </div>
    </div>
  </div>
</div>
`;
